kernel/src/device/evdev.rs

```diff
@@ -0,0 +1,623 @@
+        if events.is_empty() {
+            return Err(Error::with_message(Errno::EAGAIN, "No events available"));
+        }
+
+        // Write all events to the buffer.
+        let mut total_bytes = 0;
+        for event in events {
+            let event_bytes = event.to_bytes();
+            writer.write_fallible(&mut event_bytes.as_slice().into())?;
+            total_bytes += EVENT_SIZE;
+        }
+
+        Ok(total_bytes)
+    }
+}
+
+impl Pollable for EvdevClient {
+    fn poll(&self, mask: IoEvents, poller: Option<&mut PollHandle>) -> IoEvents {
+        self.pollee.poll_with(mask, poller, || {
+            let has_complete_packets = self.has_complete_packets();
+
+            let mut events = IoEvents::empty();
+            if has_complete_packets && mask.contains(IoEvents::IN) {
+                events |= IoEvents::IN;
+            }
+
+            events
+        })
+    }
+}
+
+impl InodeIo for EvdevClient {
+    fn read_at(
+        &self,
+        _offset: usize,
+        writer: &mut VmWriter,
+        _status_flags: StatusFlags,
+    ) -> Result<usize> {
+        let requested_bytes = writer.avail();
+        let max_events = requested_bytes / core::mem::size_of::<EvdevEvent>();
+
+        if max_events == 0 {
+            return Ok(0);
+        }
+
+        match self.process_events(max_events, writer) {
+            Ok(bytes) => Ok(bytes),
+            Err(e) if e.error() == Errno::EAGAIN => self.wait_events(IoEvents::IN, None, || {
+                self.process_events(max_events, writer)
+            }),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn write_at(
+        &self,
+        _offset: usize,
+        _reader: &mut VmReader,
+        _status_flags: StatusFlags,
+    ) -> Result<usize> {
+        // TODO: support write operation on evdev devices.
+        Err(Error::with_message(
+            Errno::ENOSYS,
+            "WRITE operation not supported on evdev devices",
+        ))
+    }
+}
+
+impl FileIo for EvdevClient {
+    fn check_seekable(&self) -> Result<()> {
+        Ok(())
+    }
+
+    fn is_offset_aware(&self) -> bool {
+        false
+    }
+
+    fn ioctl(&self, _cmd: IoctlCmd, _arg: usize) -> Result<i32> {
+        // TODO: support ioctl operation on evdev devices.
+        Err(Error::with_message(
+            Errno::EINVAL,
+            "IOCTL operation not supported on evdev devices",
+        ))
+    }
+}
+
+impl Debug for EvdevClient {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        f.debug_struct("EvdevClient")
+            .field("event_count", &self.event_count.load(Ordering::Relaxed))
+            .field("clock_type", &self.clock_type.load(Ordering::Relaxed))
+            .finish()
+    }
+}
+
+impl Drop for EvdevClient {
+    fn drop(&mut self) {
+        if let Some(evdev) = self.evdev.upgrade() {
+            evdev.detach_client();
+            evdev.close_device();
+        }
+    }
+}
+
+pub struct Evdev {
```

Comment:
To follow the convention established in the new memory devices, the type to represent device files at `/dev/input/eventX` should be named `EvdevDevice` (rather than `Evdev`) while an open file of an evdev device should be called `EvdevFile` (rather than `EvdevClient`). The two types should implement a different set of traits.